working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+example `X extends T` where `T` is nullable is neither nullable nor
+non-nullable.
+
+We say that a type `T` is **potentially nullable** if `T` is not non-nullable.
+Note that this is different from saying that `T` is nullable.  For example, a
+type variable `X extends Object?` is a type which is potentially nullable but
+not nullable.
+
+We say that a type `T` is **potentially non-nullable** if `T` is not nullable.
+Note that this is different from saying that `T` is non-nullable.  For example,
+a type variable `X extends Object?` is a type which is potentially non-nullable
+but not non-nullable.
+
+It is an error to call a method, setter, getter or operator on an expression
+whose type is potentially nullable and not `dynamic`, except for the methods,
+setters, getters, and operators on `Object`.
+
+It is an error to read a field or tear off a method from an expression whose
+type is potentially nullable and not `dynamic`, except for the methods and
+fields on `Object`.
+
+It is an error to call an expression whose type is potentially nullable and not
+`dynamic`.
+
+It is an error for an instance field with no static initializer to have a
+potentially nullable type unless the variable or field is marked with the `lazy`
+modifier.
+
+It is an error if a local variable that is potentially non-nullable and is not
+marked `lazy` is used before it is definitely assigned (see Definite Assignment
+below).
+
+It is an error if a method, function, getter, or closure with a potentially
+non-nullable return type does not definitely complete (see Definite Completion
+below).
+
+It is an error if an optional parameter (named or otherwise) with no default
+value has a potentially non-nullable type.
+
+It is an error if a required named parameter has a default value.
+
+It is an error if a named parameter that is part of a `required` group is not
+bound to an argument at a call site.
+
+It is an error to call the default `List` constructor with a length argument and
+a type argument which is potentially non-nullable.
+
+For the purposes of errors and warnings, the null aware operators `?.` and `?..`
+are checked as if the receiver of the operator had non-nullable type.
+
+It is an error for a class to extend, implement, or mixin a type of the form
+`T?` for any `T`.
+
+It is an error for a class to extend, implement, or mixin the type `Never`.
+
+It is an error to call a method, setter, or getter on a receiver of static type
+`Never` (including via a null aware operator).
+
+It is an error to apply an expression of `Never` in the function position in a
+function call.
+
+It is an error if the static type of `e` in the expression `throw e` is
+potentially nullable.
+
+It is not an error for the body of a `lazy` field to reference `this`.
+
+It is an error for a toplevel variable, a static field, or a formal parameter to
+be declared `lazy`.
+
+It is an error if the type `T` in the **on-catch** clause `on T catch` is
+potentially nullable.
```

Comment:
Not sure we can do that. It would leave some situations where you don't have a type to write because you know it's `T&Object`, but you can't write that.